README
This is a plugin for Pidgin [1] that allows you to display LaTeX [2] output into your IMs.

Important : this plugin is not compatible with gaim-encryption plugin. If you have this plugin, please be sure not having the 2 working at the same time. (for any other imcompability, please contact qjuh@users.sourceforge.net )

You need to have convert (from Image Magick [3]).

To build and install :
You can compile the plugin using
$ make
and install it with
$ make install
This will install it in ~/.purple/plugins so that only the user who install it can use it.

To install it for everybody on your computer,
$ make
$ su
# make install PREFIX="/path/to/pidgin" (this command as root user)
generally /path/to/pidgin is /usr or /usr/local. If you don't know the path then you can find out using
$ whereis pidgin
and look for the part before "/lib/pidgin".

To use
- put LaTeX code into $$ ... $$ markup, for example $$x \in \mathbb{R}$$

[1] http://www.pidgin.im/
[2] http://www.latex-project.org/
[3] http://www.imagemagick.org/

Developer :		Benjamin Moll <qjuh@users.sourceforge.net> (contact me for all infomations or suggestions)
Former Developer : 	Nicolas Schoonbroodt <nicolas@ffsa.be>
Thanks for help or patch these other developer :
		Eric Betts <bettse@onid.orst.edu>
		GRIm@ <thegrima@altern.org>
		Gof <ogoffart@kde. org>
		Nicolai Stange (nic-stange@t-online.de)


CHANGELOG
*************
Version 1.2
*************
	- added support for LaTeX in Chats like IRC
	- minor bugs fixed

*************
Version 1.1
*************
	- fix STILL_ACTIVE bug on windows
	- some minor bugs fixed

*************
Version 1.0
*************
	- changes for the new API of Pidgin applied
	- added getfilename()
	- changed tempfile-generation using purple-function
	- some more blacklist-checking

*************
Version 0.4
*************
        - windows port
	- < 1024 bytes image bug corrected


*************
Version 0.3
*************
	- no more need of tex2im
	- fix crash bug when no image were created
	- fix gaim-encryption bug
	- now use $$ markup (compatibility with KopeteTeX) 
		old markup ([tex] ... [/tex]) will pobably be erased
		later, so please use new system (if you have any
		objection, speak from that with me


*************
version 0.2
*************
	- Work with gaim v1.0.3 
	- The logs of the conversation now log the original message with the
	  [tex]...[/tex] markup
	- Now compile easier with a Makefile
	
***********
Version 0.1
***********
	- Initial version
	- site http://sourceforge.net/projects/gaim-latex/


LaTeX.h
/*
 * LaTeX.c
 * pidgin-latex plugin
 *
 * This a plugin for Pidgin to display LaTeX formula in conversation
 *
 * PLEASE, send any comment, bug report, etc. to the trackers at sourceforge.net
 *
 * Copyright (C) 2006-2007 Benjamin Moll (qjuh@users.sourceforge.net)
 * some portions : Copyright (C) 2004-2006 Nicolas Schoonbroodt (nicolas@ffsa.be)
 *                 Copyright (C) 2004-2006 GRIm@ (thegrima@altern.org).
 * 		   Copyright (C) 2004-2006 Eric Betts (bettse@onid.orst.edu).
 * Windows port  : Copyright (C) 2005-2006 Nicolai Stange (nic-stange@t-online.de)
 * Other portions heavily inspired and copied from gaim sources
 * Copyright (C) 1998-2007 Pidgin developers pidgin.im
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; This document is under the scope of
 * the version 2 of the License, or any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program (see COPYING); if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 */
#ifndef _PIDGIN_LATEX_H_
#define _PIDGIN_LATEX_H_
#ifndef G_GNUC_NULL_TERMINATED
#  if __GNUC__ >= 4
#    define G_GNUC_NULL_TERMINATED __attribute__((__sentinel__))
#  else
#    define G_GNUC_NULL_TERMINATED
#  endif /* __GNUC__ >= 4 */
#endif /* G_GNUC_NULL_TERMINATED */

#define PURPLE_PLUGINS

//include
#include <libpurple/conversation.h>
//#include <libpurple/core.h>
#include <libpurple/debug.h>
#include <libpurple/signals.h>
#include <libpurple/imgstore.h>
#include <libpurple/util.h>
#include <libpurple/notify.h>
#include <libpurple/server.h>
#include <libpurple/log.h>
#include <libpurple/version.h>

// Constant
#define IMG_BEGIN "<img id=\""
#define IMG_END "\">"

#define BEG "[tex]"
#define END "[/tex]"
#define KOPETE_TEX "$$"

#define LATEX_PLUGIN_ID "qjuh-LaTeX"
#define WEBSITE "http://sourceforge.net/projects/pidgin-latex/"

#define HEADER "\\documentclass[12pt]{article}\\usepackage[dvips]{graphicx}\\usepackage{amsmath}\\usepackage{amssymb}\\pagestyle{empty}"
#define HEADER_MATH "\\begin{document}\\begin{gather*}"

#define FOOTER "\\end{document}"
#define FOOTER_MATH "\\end{gather*}"

#define FILTER_AND "&amp;"
#define FILTER_LT "&lt;"
#define FILTER_GT "&gt;"
#define FILTER_BR "<br>"

/* Yes, this is simply a copy/paste of KopeteTex blacklist. */
/* But too bad in LaTeX and system security to verify all   */
/* of this */
#define NB_BLACKLIST (42)
#define BLACKLIST {"\\def","\\let","\\futurelet","\\newcommand","\\renewcomment","\\else","\\fi","\\write","\\input","\\include","\\chardef","\\catcode","\\makeatletter","\\noexpand","\\toksdef","\\every","\\errhelp","\\errorstopmode","\\scrollmode","\\nonstopmode","\\batchmode","\\read","\\csname","\\newhelp","\\relax","\\afterground","\\afterassignment","\\expandafter","\\noexpand","\\special","\\command","\\loop","\\repeat","\\toks","\\output","\\line","\\mathcode","\\name","\\item","\\section","\\mbox","\\DeclareRobustCommand"}


// prototypes

/* Verify Blacklist */
/* return true if one word of the message is blacklisted */
static gboolean is_blacklisted(char *message);

/* 
 * latex_to_image creates PNG-image  with the LaTeX code pointed by *latex
 * *latex points to latex-input
 * **file_tex receives filename of temporary generated LaTeX-file; file must be deleted and string must be freed by caller
 * **file_dvi receives filename of temporary generated dvi-file; file must be deleted and string must be freed by caller
 * **file_ps receives filename of temporary generated ps-file; file must be deleted and string must be freed by caller
 * **file_png receives filename of temporary generated png-file; file must be deleted and string must be freed by caller
 *
 * returns TRUE on success, false otherwise
 */
static gboolean latex_to_image(char *latex, char **file_tex, char **file_dvi, char **file_ps, char **file_png);

/*
 * Transform *tmp2 extracting some *startdelim here *enddelim thing, make png-image from latex-input
 *  and tmp2 becomes 'some<img="number">thing'
 * returns TRUE on success, FALSE otherwise
 */
static gboolean analyse(char **tmp2, char *startdelim, char *enddelim, gboolean remote);

/*
 * pidgin_latex_write perform the effective write of the latex code in the IM windows
 * 	*conv is a pointer onto the conversation context
 *	*nom is the name of the correspondent
 *	*message is the modified message with the image
 *	*messFlags is Flags related to the messages
 *	*original is the original message unmodified
 * return TRUE.
 */
static gboolean pidgin_latex_write(PurpleConversation *conv, const char *nom, char *message, PurpleMessageFlags messFlag, char *original);

/* to intercept in- and outgoing messages */
static gboolean message_send(PurpleAccount *account, const char *who, char **buffer, PurpleConversation *conv, PurpleMessageFlags flags);

/*
 * getdirname returns the directory's part of a filename.
 * Parsing is done OS-dependently (with path-separator as defined in glib/gutils.h)
 *      *path is the filename you want to extract a directory-part from
 * return directory's part on success, NULL otherwise; must be freed with
 * free()!
 */
static char* getdirname(const char const *file);

/*
 * get_latex_cmd returns a system()-executable-command for latex
 *
 * return the executable command if successfull, NULL otherwise; must be
 * freed with free()
 */
static char* get_latex_cmd();

/*
 * get_dvips_cmd returns a system()-executable-command for dvips
 *
 * return the executable command if successfull, NULL otherwise; must be
 * freed with free()
 */
static char* get_dvips_cmd();

/*
 * get_convert_cmd returns a system()-executable-command for convert
 *
 * return the executable command if successfull, NULL otherwise; must be
 * freed with free()
 */
static char* get_convert_cmd();

/*
 * win32_purple_notify extracts error information of last occured WIN32-Error due to an API-call and asserts it to gaim via purple_notify_error
 * Error message will be prepended by *prep:
 */
void win32_purple_notify_error(char *prep);

/*
 * searchPATH searches the PATH-environment for the specified file file.
 * *file is the name of the executable to search for
 * returns the right full path, e.g. with the executable's name appended, NULL on failure, must be freed with free()
 */
char* searchPATH(const char const *file);

/*
 * execute executes the *cmd with opts appended WITHOUT ANY SPACES INBETWEEN to the commandline.
 * Advantage to system()-call under win32 ist that no console window pops up.
 * On systems other than windows this function is just a wrapper around the system()-call.
 * *cmd is the comannd to execute
 * *opts[] is an array of elements to be appended to the commandline
 * copts is the count of elements within *opts[]
 *
 * returns -1 if execution failed, otherwise the return code of the executed program.
*/
#ifdef _WIN32
static LPDWORD execute(char *cmd, char *opts[], int copts);
#else
static int execute(char *cmd, char *opts[], int copts);
#endif


#endif


LaTeX.c
/*
 * LaTeX.c
 * pidgin-latex plugin
 *
 * This a plugin for Pidgin to display LaTeX formula in conversation
 *
 * PLEASE, send any comment, bug report, etc. to the trackers at sourceforge.net
 *
 * Copyright (C) 2006-2007 Benjamin Moll (qjuh@users.sourceforge.net)
 * some portions : Copyright (C) 2004-2006 Nicolas Schoonbroodt (nicolas@ffsa.be)
 *                 Copyright (C) 2004-2006 GRIm@ (thegrima@altern.org).
 * 		   Copyright (C) 2004-2006 Eric Betts (bettse@onid.orst.edu).
 * Windows port  : Copyright (C) 2005-2006 Nicolai Stange (nic-stange@t-online.de)
 * Other portions heavily inspired and copied from gaim sources
 * Copyright (C) 1998-2007 Pidgin developers pidgin.im
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; This document is under the scope of
 * the version 2 of the License, or any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program (see COPYING); if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 */
#include "LaTeX.h"

#include <stdio.h>
#include <string.h>
#include <unistd.h>

#ifndef MAX
#define MAX(a,b) ( (a) > (b) ? (a ): (b))
#endif

#ifdef _WIN32
#include <windows.h>
#endif

static void
open_log(PurpleConversation *conv)
{
	conv->logs = g_list_append(NULL, purple_log_new(conv->type == PURPLE_CONV_TYPE_CHAT ? PURPLE_LOG_CHAT :
							   PURPLE_LOG_IM, conv->name, conv->account,
							   conv, time(NULL), NULL));
}

#ifdef _WIN32
void win32_purple_notify_error(char *prep)
{
  char *errmsg=NULL;
  char *finalmsg=NULL;
  if(!FormatMessage(FORMAT_MESSAGE_ALLOCATE_BUFFER | FORMAT_MESSAGE_FROM_SYSTEM, NULL, GetLastError(), MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT), (void*)&errmsg, 0, NULL))
  {
      purple_notify_error(NULL, "LaTeX" ,"Can't display error message.", NULL);
      return;
  }

  if(prep)
  {
    finalmsg=malloc((strlen(errmsg)+ strlen(prep) + 3)*sizeof(char));
    if(!finalmsg)
    {
      purple_notify_error(NULL, "LaTeX", "Can't display error message.", NULL);
      LocalFree(errmsg); /* we can't do anything more for you */
      return;
    }
    strcpy(finalmsg, prep);
    strcat(finalmsg, ": ");
    strcat(finalmsg, errmsg);
    LocalFree(errmsg);
  }
  else
  {
    finalmsg = malloc((strlen(errmsg)+1)*sizeof(char));
    if(!finalmsg)
    {
      purple_notify_error(NULL, "LaTeX", "Can't display error message.", NULL);
      LocalFree(errmsg); /* we can't do anything more for you */
      return;
    }
    strcpy(finalmsg, errmsg);
    LocalFree(errmsg);
  }
  purple_notify_error(NULL, "LaTeX", finalmsg, NULL);
  free(finalmsg);
  return;
}
#endif

static char* getdirname(const char const *file)
{
  char *s=NULL;
  char *r=NULL;
  s=strrchr(file, G_DIR_SEPARATOR);
  if(!s) /* just a pure filename without dir? */
  {
    /* Here is no standard-, but GNU-bahaviour of getcwd assumed.
       Note that msdn.microsoft.com defines the same as GNU.
     */
    return getcwd(NULL,0);
  }

  s+=1; /* get the G_DIR_SEPARATOR at the end of directory-string */
  r=malloc(s-file+sizeof(char));
  if(r)
  {
    memcpy(r,file, s-file);
    r[(s-file)/sizeof(char)]='\0';
  }
  return r;
}

static char* getfilename(const char const *file)
{
  char *s=NULL;
  char *r=NULL;
  s=strrchr(file,G_DIR_SEPARATOR);
  if(!s)
  {
    r=malloc((strlen(file)+1)*sizeof(char));
    strcpy(r,file);
    return r;
  }

  s+=1;
  r=malloc((strlen(file)+1)*sizeof(char)+file-s);
  if(r)
  {
    memcpy(r,s,strlen(file)*sizeof(char)+file-s);
    r[strlen(file)+(file-s)/sizeof(char)]='\0';
  }
  return r;
}

char* searchPATH(const char const *file)
{
  char *cmd=NULL;
#ifdef _WIN32
  DWORD sz=0;
  DWORD sz2=0;

  sz=SearchPath(NULL, file, ".exe", 0, cmd, NULL);
  cmd = malloc((sz+1)*sizeof(TCHAR));
  if(cmd)
  {
    sz2=SearchPath(NULL, file, ".exe", sz+1, cmd, NULL);
    if(!sz2)
    {
      free(cmd);
      cmd=NULL;
    }
  }
#else
  cmd=malloc((strlen(file)+1)*sizeof(char));
  if(cmd)
    strcpy(cmd, file);
#endif

  return cmd;
}

#ifdef _WIN32 /* we could take system(), too, but this opens ugly console-windows within IM-Session*/
static LPDWORD execute(char *cmd, char *opts[], int copts)
{
  int i=0;
  int len=strlen(cmd) + 4;
  char *params=NULL;
  LPDWORD exitcode=0;
  for(i=0; i<copts; ++i)
    len+=(strlen(opts[i]))*sizeof(char);

  params=malloc(len);
  if(!params)
    return -1;

  strcpy(params, "\"");
  strcat(params, cmd);
  strcat(params, "\" ");

  for(i=0; i<copts; ++i)
    strcat(params, opts[i]);

  STARTUPINFO sup;
  PROCESS_INFORMATION pi;
  ZeroMemory( &sup, sizeof(sup) );
  sup.cb = sizeof(sup);
  ZeroMemory( &pi, sizeof(pi) );
  sup.wShowWindow = SW_HIDE;
  sup.dwFlags = STARTF_USESHOWWINDOW;

  if(!CreateProcess(NULL, params, NULL, NULL, TRUE, 0, NULL, NULL, &sup, &pi))
  {
    win32_purple_notify_error(cmd);
    free(params);
    return -1;
  }

  free(params);
  if(WAIT_OBJECT_0!=WaitForSingleObjectEx(pi.hProcess, INFINITE, FALSE))
  {
    win32_purple_notify_error(cmd);
    CloseHandle(pi.hProcess);
    CloseHandle(pi.hThread);
    return -1;
  }

  do {
    if(!GetExitCodeThread(pi.hThread, &exitcode))
    {
      win32_purple_notify_error(cmd);
      return -1;
    }
    Sleep(10);
  } while(exitcode==STILL_ACTIVE);
  CloseHandle(pi.hProcess);
  CloseHandle(pi.hThread);
  return exitcode;
}
#else /* not windows, take system() */
static int execute(char *cmd, char *opts[], int copts)
{
  int i=0;
  int exitcode=-1;
  int len=strlen(cmd) + 4;
  char *params=NULL;

  for(i=0; i<copts; ++i)
    len+=(strlen(opts[i]))*sizeof(char);

  params=malloc(len);
  if(!params)
    return -1;

  strcpy(params, "\"");
  strcat(params, cmd);
  strcat(params, "\" ");

  for(i=0; i<copts; ++i)
    strcat(params, opts[i]);

  exitcode=system(params);
  free(params);
  return exitcode;
}
#endif

static char* get_latex_cmd()
{
  return searchPATH("latex");
}

static char* get_dvips_cmd()
{
  return searchPATH("dvips");
}

static char* get_convert_cmd()
{
#ifdef _WIN32
    /*open registry*/
    DWORD type;
    DWORD cbData;
    char *pData=NULL;
    HKEY hk=NULL;
    RegOpenKeyEx(HKEY_LOCAL_MACHINE, "SOFTWARE\\ImageMagick\\Current",0,KEY_READ,&hk);
    RegQueryValueEx(hk, "BinPath", NULL, &type, NULL, &cbData);
    cbData=cbData + (1+12)*sizeof(char);
    pData=malloc(cbData);
    if(pData && RegQueryValueEx(hk, "BinPath", NULL, &type, pData, &cbData)==ERROR_SUCCESS)
    {
      CloseHandle(hk);
      strcat(pData,"\\convert.exe");
      return pData;
    }

    CloseHandle(hk);
    free(pData);
    return searchPATH("convert");
#else
  return searchPATH("convert");
#endif
}

static gboolean is_blacklisted(char *message)
{
  char *not_secure[NB_BLACKLIST] = BLACKLIST;
  int i;
  for (i = 0 ; i < NB_BLACKLIST ; i++)
    {
      char *begin_not_secure = malloc((strlen(not_secure[i])+9)*sizeof(char));
      strcpy(begin_not_secure,"\\begin{");
      strcat(begin_not_secure,not_secure[i]+0x01);
      strcat(begin_not_secure,"}");
      if (strstr(message, not_secure[i]) != NULL || strstr(message, begin_not_secure)) return TRUE;
    }
  return FALSE;
}

static gboolean latex_to_image(char *latex, char **file_tex, char **file_dvi, char **file_ps, char **file_png)
{
  FILE * texfile;
  char *file_tmp = NULL;
  char *tmpdir = NULL;
  char *cmdlatex = NULL;
  char *cmddvips = NULL;
  char *cmdconvert = NULL;

  /* the following is new and related to temporary-filename-generation */
  texfile = purple_mkstemp(&file_tmp,TRUE);
  *file_tex=malloc((strlen(file_tmp)+5)*sizeof(char));
  *file_dvi=malloc((strlen(file_tmp)+5)*sizeof(char));
  *file_ps=malloc((strlen(file_tmp)+4)*sizeof(char));
  *file_png=malloc((strlen(file_tmp)+5)*sizeof(char));
  if(!(file_tmp && *file_tex && *file_dvi && *file_ps && *file_png))
  {
    fclose(texfile);
    unlink(file_tmp);
    free(file_tmp);
    free(*file_tex);
    free(*file_dvi);
    free(*file_ps);
    free(*file_png);
    *file_tex=*file_dvi=*file_ps=*file_png=NULL;
    return FALSE;
  }
  strcpy(*file_tex, file_tmp);
  strcat(*file_tex, ".tex");
  strcpy(*file_dvi, file_tmp);
  strcat(*file_dvi, ".dvi");
  strcpy(*file_ps, file_tmp);
  strcat(*file_ps, ".ps");
  strcpy(*file_png, file_tmp);
  strcat(*file_png, ".jpg");
  unlink(file_tmp);
  free(file_tmp);
  fclose(texfile);

  if (! (texfile = fopen(*file_tex, "w"))) return FALSE;
  fprintf(texfile, HEADER HEADER_MATH "%s" FOOTER_MATH FOOTER, latex);
  fclose (texfile);

  tmpdir=getdirname(*file_tex);
  /* generate commands, also new */
  char *latexopts[5]={"--interaction=nonstopmode", " ", "\"", *file_tex, "\""};
  char *dvipsopts[10]={"-E", " ", "-x 1300", " ", "-o", " \"", *file_ps, "\" \"", *file_dvi, "\""};
  char *convertopts[5]={"\"", *file_ps, "\" \"", *file_png, "\""};
  cmdlatex=get_latex_cmd();
  cmddvips=get_dvips_cmd();
  cmdconvert=get_convert_cmd();

  if (!tmpdir || chdir(tmpdir) || !cmdlatex || !cmddvips || !cmdconvert)
  {
    free(*file_dvi);
    free(*file_ps);
    free(*file_png);
    *file_dvi=*file_ps=*file_png=NULL;
    free(cmdlatex);
    free(cmddvips);
    free(cmdconvert);
    return FALSE;
  }

  free(tmpdir);
  if((execute(cmdlatex, latexopts, 5) || execute(cmddvips, dvipsopts, 10) ||  execute(cmdconvert, convertopts, 5))) return FALSE;

  free(cmdlatex);
  free(cmddvips);
  free(cmdconvert);

  return TRUE;
}

static gboolean analyse(char **tmp2, char *startdelim, char *enddelim, gboolean remote)
{
  int pos1, pos2, idimg;
  char *ptr1, *ptr2;
  char *file_tex=NULL;
  char *file_dvi=NULL;
  char *file_ps=NULL;
  char *file_png=NULL;

  ptr1 = strstr(*tmp2, startdelim);
  while(ptr1 != NULL)
    {
      char *tex, *message, *filter, *idstring;
      gchar *name, *buf, *filedata;
      size_t size;
      GError *error = NULL;

      pos1 = strlen(*tmp2) - strlen(ptr1);

      // Have to ignore the first 2 char ("$$") --> & [2]
      ptr2 = strstr(&ptr1[strlen(startdelim)], enddelim);
      if (ptr2 == NULL)
	{ return FALSE; }

      pos2 = strlen(*tmp2) - strlen(ptr2) + strlen(enddelim);

      if ((tex = malloc(pos2 - pos1 - strlen(enddelim) - strlen(startdelim) + 1)) == NULL)
	{
	  // TODO: Report the error
	  return FALSE;
	}

      strncpy(tex, &ptr1[strlen(startdelim)], pos2 - pos1 - strlen(startdelim)-strlen(enddelim));
      tex[pos2-pos1-strlen(startdelim)-strlen(enddelim)] = '\0';

      // Pidgin transforms & to &amp; and I make the inverse transformation
      while ( (filter = strstr(tex, FILTER_AND) ) != NULL)
	{
	  strcpy(&tex[strlen(tex) - strlen(filter) + 1], &filter[5]);
	}
      // Pidgin transforms < to &lt
      while ( (filter = strstr(tex, FILTER_LT) ) != NULL)
	{
	  strcpy(&tex[strlen(tex) - strlen(filter)], "<");
	  strcpy(&tex[strlen(tex) - strlen(filter)] + 1, &filter[4]);
	}
      // Pidgin transforms > to &gt
      while ( (filter = strstr(tex, FILTER_GT) ) != NULL)
	{
	  strcpy(&tex[strlen(tex) - strlen(filter)], ">");
	  strcpy(&tex[strlen(tex) - strlen(filter)] + 1, &filter[4]);
	}
      // <br> filter
      while ( (filter = strstr(tex, FILTER_BR) ) != NULL)
	{
	  strcpy(&tex[strlen(tex) - strlen(filter)], &filter[4]);
	}
      
      // Creates the image in file_png
      if (!latex_to_image(tex, &file_tex, &file_dvi, &file_ps, &file_png)) { free(tex); return FALSE; };
      free(tex);

      // loading image
       if (!g_file_get_contents(file_png, &filedata, &size, &error))
	{
	  purple_notify_error(NULL, "LaTeX", error->message, NULL);
	  g_error_free(error);
	  return FALSE;
	}

      unlink(file_tex);
      file_tex[strlen(file_tex)-4]='\0';

      strcat(file_tex, ".aux");
      unlink(file_tex);
      file_tex[strlen(file_tex)-4]='\0';
      strcat(file_tex, ".log");
      unlink(file_tex);
      unlink(file_dvi);
      unlink(file_ps);
      unlink(file_png);

      free(file_tex);
      free(file_dvi);
      free(file_ps);
      free(file_png);

      
      name = "pidginTeX.jpg";

      idimg = purple_imgstore_add_with_id(filedata, MAX(1024,size), name);
      filedata = NULL;

      if (idimg == 0)
	{
	  buf = g_strdup_printf("Failed to store image.");
	  purple_notify_error(NULL,"LaTeX", buf, NULL);
	  g_free(buf);
	  return FALSE;
	}

      idstring = malloc(10);
      sprintf(idstring, "%d", idimg);

      // making new message
      if ((message = malloc (strlen(*tmp2) - pos2 + pos1 + strlen(idstring) + strlen(IMG_BEGIN) + strlen(IMG_END) + 1)) == NULL)
	{
	  purple_notify_error(NULL,"LaTeX", "couldn't make the message.", NULL);
	  return FALSE;
	}

      if (pos1 > 0)
	{
	  strncpy(message, *tmp2, pos1);
	  message[pos1] = '\0';
	  strcat(message, IMG_BEGIN);
	} else  {
	  strcpy(message, IMG_BEGIN);
	}
      strcat(message, idstring);
      strcat(message, IMG_END);

      free(idstring);

      if (pos2 < strlen(*tmp2))
      { strcat(message, &ptr2[strlen(enddelim)]); }

      free(*tmp2);
      if ((*tmp2 = malloc(strlen(message)+1)) == NULL)
      {
	purple_notify_error(NULL,"LaTeX", "couldn't split the message.", NULL);
	return FALSE;
      }

      strcpy(*tmp2, message);
      free(message);

      ptr1 = strstr(*tmp2, startdelim);
    }
  return TRUE;
}

static gboolean pidgin_latex_write(PurpleConversation *conv, const char *nom, char *message, PurpleMessageFlags messFlag, char *original)
{
  gboolean logflag;

  // writing log
  logflag = purple_conversation_is_logging(conv);

  if (logflag)
    {
      GList *log;

      if (conv->logs == NULL)
        open_log(conv);

      log = conv->logs;
      while (log != NULL) {
        purple_log_write((PurpleLog *)log->data, messFlag, nom, time(NULL), original);
        log = log->next;
      }
      purple_conversation_set_logging(conv,FALSE);
    }
  
  if(purple_conversation_get_type(conv) == PURPLE_CONV_TYPE_CHAT)
    purple_conv_chat_write(PURPLE_CONV_CHAT(conv), nom, message, messFlag, time(NULL));
  else if(purple_conversation_get_type(conv) == PURPLE_CONV_TYPE_IM)
    purple_conv_im_write(PURPLE_CONV_IM(conv), nom, message, messFlag, time(NULL));
 
  if (logflag)
    purple_conversation_set_logging(conv,TRUE);

  return FALSE;
}

static gboolean message_send(PurpleAccount *account, const char *who, char **buffer, PurpleConversation *conv, PurpleMessageFlags flags)
{
  char *tmp2;
  int t[10], i = 0;

  // if nothing to do
  if (strstr(*buffer,KOPETE_TEX) == NULL)
    {
      return FALSE;
    };

  if (is_blacklisted(*buffer)) return FALSE;

  if((tmp2 = malloc(strlen(*buffer)+1)) == NULL)
    {
      // TODO: Notify Error
      return FALSE;
    }

  strcpy(tmp2,*buffer);

  if (analyse(&tmp2, KOPETE_TEX, KOPETE_TEX, FALSE) )
    {
      pidgin_latex_write(conv, who, tmp2, flags, *buffer);

      free(tmp2);
      return TRUE;
    }

  free(tmp2);

  return FALSE;
}

static gboolean plugin_load(PurplePlugin *plugin)
{
  void *conv_handle = purple_conversations_get_handle();

  purple_signal_connect(conv_handle, "writing-im-msg",
		      plugin, PURPLE_CALLBACK(message_send), NULL);

  purple_signal_connect(conv_handle, "writing-chat-msg",
		      plugin, PURPLE_CALLBACK(message_send), NULL);

  purple_debug(PURPLE_DEBUG_INFO, "LaTeX", "LaTeX loaded\n");

  return TRUE;
}

static gboolean plugin_unload(PurplePlugin * plugin)
{
  void *conv_handle = purple_conversations_get_handle();

  purple_signal_disconnect(conv_handle, "writing-im-msg", plugin, PURPLE_CALLBACK(message_send));
  purple_signal_disconnect(conv_handle, "writing-chat-mg", plugin, PURPLE_CALLBACK(message_send));

  return TRUE;
}


static PurplePluginInfo info =
  {
    PURPLE_PLUGIN_MAGIC,
    PURPLE_MAJOR_VERSION,
    PURPLE_MINOR_VERSION,
    PURPLE_PLUGIN_STANDARD,                             /**< type           */
    NULL,                                             /**< ui_requirement */
    0,                                                /**< flags          */
    NULL,                                             /**< dependencies   */
    PURPLE_PRIORITY_DEFAULT,                            /**< priority       */

    LATEX_PLUGIN_ID,                                  /**< id             */
    "LaTeX",                                      /**< name           */
    "1.2",                                        /**< version        */
    /**  summary        */
    "To display LaTeX formula into Pidgin conversation.",
    /**  description    */
    "Put LaTeX-code between $$ ... $$ markup to have it displayed as Picture in your conversation.\nRemember that your contact needs an similar plugin or else he will just see the pure LaTeX-code\nYou must have LaTeX and ImageMagick installed (in your PATH)",
    "Benjamin Moll <qjuh@users.sourceforge.net>\nNicolas Schoonbroodt <nicolas@ffsa.be>\nNicolai Stange <nic-stange@t-online.de>",   /**< author       */
    WEBSITE,                                          /**< homepage       */
    plugin_load,                                      /**< load           */
    plugin_unload,                                    /**< unload         */
    NULL,                                             /**< destroy        */
    NULL,                                             /**< ui_info        */
    NULL,                                             /**< extra_info     */
    NULL,
    NULL,
    NULL,
    NULL,
    NULL,
    NULL
  };



static void
init_plugin(PurplePlugin *plugin)
{
}

PURPLE_INIT_PLUGIN(LaTeX, init_plugin, info)
